lib/tree.rb

```diff
@@ -16,51 +16,163 @@ def initialize
+    else 
+      find_helper(current.right, key)
+    end
+  end 
+
+  # Time Complexity: O(n) => because we traverse every node
+  # Space Complexity: balanced tree: O(log n) & unbalanced tree: O(n) => because of the System Stack
+  # In-Order: Left, Current, Right
+
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # recursive solution
+  def inorder_helper(current, list)
+    return list if current.nil?
+
+    # left -> root -> right
+    inorder_helper(current.left, list)
+    list << {:key => current.key, :value => current.value}
+    inorder_helper(current.right, list)
+
+    return list
+  end
+
+  # Time Complexity: O(n) => because we traverse every node
+  # Space Complexity: balanced tree: O(log n) & unbalanced tree: O(n) => because of the System Stack
+  # Pre-Order: Current, Left, Right
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
+  end
+
+  # recursive solution
+  def preorder_helper(current, list)
+    return list if current.nil?
+
+    # root ->  Left -> Right
+    list << {:key => current.key, :value => current.value}
+    preorder_helper(current.left, list)
+    preorder_helper(current.right, list)
+
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) => because we traverse every node
+  # Space Complexity: balanced tree: O(log n) & unbalanced tree: O(n) => because of the System Stack
+  # Post-Order: Left, Right, Current
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
+  end
+
+  # recursive solution
+  def postorder_helper(current, list)
+    return list if current.nil?
+
+    # Left -> Right -> root
+    postorder_helper(current.left, list)
+    postorder_helper(current.right, list)
+    list << {:key => current.key, :value => current.value}
+
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) => because we traverse every node
+  # Space Complexity: balanced tree: O(log n) & unbalanced tree: O(n) => because of the System Stack
   def height
-    raise NotImplementedError
+    return height_helper(@root, 0)
+  end
+
+  # recursive solution
+  def height_helper(current, count)
+    return count if current.nil?
+
+    count += 1
+
+    current_left = height_helper(current.left, count)
+    current_right = height_helper(current.right, count)
+
+    return [current_left, current_right].max
   end
 
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(V + E) => Since every vertex and every edge will be checked once
+  # Space Complexity: O(V) => reason?
+
+  # (note) Breadth First Traversals which visit each node level, by level and Depth First Traversals which visit a node's children before it's siblings.
+  # In a BFS, you first explore all the nodes one step away, then all the nodes two steps away, etc
   def bfs
```

Comment:
👍 , actually the space complexity is O(n) because you're building an array of all the nodes.  The O(V) refers to graphs, or the number of nodes in the graph.